float32 rand1
float32 rand2
